src/fileformat/pfsoutyuv.1
.TH "pfsoutyuv" 1
.SH NAME
pfsoutyuv \- Write frames to .yuv file used by many video codecs
.SH SYNOPSIS
.B pfsoutyuv [--verbose] [--quiet] [--bitdepth] [--colorspace] [--downsample-filter] [--chroma-format] 
<file> [<file>...]

.SH DESCRIPTION
Use this command to write the pfsstream to an uncompressed .yuv file.
.PP
If the output color space is LDR (bt709), the input pfs stream should be in the gamma corrected sRGB color space. If the output color space is HDR (pq2020 or hgl2020), the input pfs stream should be in linear RGB color space. 
.SH OPTIONS
.TP
.B \--bitdepth, -b
bit-depth of the integer values stored in the yuv file. a single color component is stored as a byte if bit-depth is 8, or as two bytes otherwise. default: 10 [bits]
.TP
.B \--colorspace, -c
Currently it can be either: 
.IP
.I pq2020
for HDR images with BT.2020 colorspace and PQ transfer function, or 
.IP
.I hlg2020
for HDR images with BT.2020 colorspace and HLG transfer function, or 
.IP
.I bt709
for LDR images in BT.709 (rec.709) colorspace without any transfer function.
.PP
If no option is specified, colorspace is inferred from the LUMINANCE tag in the pfsstream: pq2020 if LUMINANCE is ABSOLUTE or RELATIVE, bt709 if LUMINANCE is DISPLAY.
.TP
.B \--chroma-format, -f
The sampling format for the chroma channels. Can be one of:
.IP
.I 444
The standard 4:4:4 format with no chroma downsampling.
.IP
.I 420
The 4:2:0 format with both colour channels being downsampled by half.
.IP 
default: 420
.TP
.B \--downsample-filter, -f
The filter weights used when downsampling to 4:2:0. Can be either
.IP
.I f0
Weighting = [1/8, 3/4, 1/8]
.IP
.I f1
Weighting = [1/4, 1/2, 1/4]
.IP
default: f0
.TP
.B \--srgb-input -s 
Set if the input is gamma corrected, though this will be inferred otherwise from the LUMINANCE tag in the pfs stream
.TP
.B \--linear-input -l
As above, set if the input contains linear luminance values
.TP
.B \--verbose, -v
Print extra information to stderr.
.TP
.B \--quiet, -q
Disable warning message about the change in handling absolute values in RGBE files. See IMPORTANT NOTE below.

.SH EXAMPLES

.TP
 pfsin frame%04d.exr | pfsoutyuv video%s.yuv
Converts a sequence of frames frame0000.exr, frame0001.exr, ... to a yuv file. 

.SH "SEE ALSO"
.BR pfsin (1)

.SH IMPORTANT NOTE

.SH BUGS
Please report bugs and comments on implementation to 
the discussion group http://groups.google.com/group/pfstools


ChangeLog
pfstools 2.2.0 <12.08.2021>
	* Added: v210 format in pfsinyuv
	* Fixed: memleak in pfssize
	* Fixed: Fix format-security errors with Octave 5.1 (thanks to Orion Poplawski)
	* Added: pfs_automerge
	* Fixed: typos in the documentation (thanks to Andreas Metzler)
	* Added: simple deghosting in pfshdrcalibrate (-d option)
	* Fixed: pfsalign now uses free AKAZE feature detector, avoiding issues with missing non-free OpenCV classes
	* Fixed: Updated installation instruction for Ubuntu 18.04
	* Fixed: Fixed a number of compiler warnings from g++ 7.4.0
	* Fixed: Octave interface upgraded to Octave-6 (thanks to Michal)
	* Fixed: pfstmo_durand02 - reversed the old fix that made images over-saturated. Now 99.5th percentile is mapped to white. 
	* Removed: pfsinjpeghdr, pfsoutjpeghdr (discontinued)
	* Fixed: compilation fails with GCC11/std=c++17, duplicate "clamp" definition (thanks to Stefan)
	* Added: pfshdrcalibrate now performs exposure merging in a noise-optimal manner
	* Added: pfsindcraw now calls libraw's dcraw_emu instead of dcraw if the former is available

pfstools 2.1.0 <13.11.2017>
	* Added: pfsinyuv/pfsoutyuv to handle uncompressed video .yuy
	* Added: Support for PQ2020 and HLG2020 color spaces (in pfs*yuv)
	* Added: pfsview now builds with Qt5 instead of Qt4
	* Fixed: pfsglview crashing on OSX (thanks to Michael)
	* Fixed: issues with compiling matlab MEX functions on Ubuntu
	* Added: installation instruction for Ubuntu (README.Ubuntu)
	* Added: experimemtal version of deghosting in pfshdrcalibrate (poor performance)

pfstools 2.0.6 <15.03.2017>
	* Fixed: --frame frame range spec can now handle negative frame order, e.g. 30:-1:0
	* Added: Added photon-noise weighting in pfshdrcalibrate - very small improvement for some images
	* Fixed: pfsview does not change zoom settings when flipping between next/previous images
	* Fixed: Fixed bug in pfsin/outimgmagics - missing call to InitializeMagick causing assertion fault
	* Fixed: Fixed bug in pfsouthdrhtml - missing call to InitializeMagick causing assertion fault
	* Added: Matlab MEX files can now be compiled on Windows from Cygwin using VisualStudio compiler

pfstools 2.0.5 <26.05.2016>
	* fixed: pfsinppm now can read multiple frames using --frames but also from a ppm file
	* fixed: a few typos in the documentation and code (thanks to Andreas)
	* fixed: fixed octave interface to work with Octave 3.8.1 (and hopefully later versions)
	* fixed: octave-config used to automatically determine were to install Octave files
	* fixed: compilation issues with c++-11
	* fixed: bug in the assert statement made some matlab mex functions to crash (in debug mode)
	* fixed: version number string is now correctly reported
	* fixed: reverts mistakenly introduced fix from 2.0.0: pfsinpfm and pfsoutpfm do not flip images (top-bottom)
	* fixed: pfs handles now very long tags (by truncating them)

pfstools 2.0.4 <15.07.2015>
	* fixed: added installation of octave-based scripts: pfsoctavelum pfsoctavergb pfsstat
	* fixed: libraries installed in lib64 if needed (thanks to Orion for the patch)
	* fixed: added "so" version to the pfs.so library (thanks to Orion for the patch)
	* fixed: Replaced depreciated OctaveMap for compatibility with octave 4.0.0

pfstools 2.0.3 <28.04.2015>
	* fixed: permission issues in the cmake scripts
	* fixed: Out-of-source compilation issues in octave (thanks to Tomas)
	* added: pfsoutexr has new --float32 option to store color as 32-bit float
	* upated: pfsoutexr assumes --fix-halfmax as default and has additional switch --clamp-halfmax to disable this behavior
	* fixed: pfshdrcalibrate: serious bug that caused Robertson's method to fail most of the time
	* fixed: pfshdrcalibrate: Improved clipping of saturated pixels should avoid most of the artefacts
	* added: jpeg2hdrgen - now the command can use "identify" from ImageMagick instead of "jhead"

pfstools 2.0.2 <08.03.2015>
	* fixed: Fixed build and install scripts for hdrhtml
	* fixed: Fixed building FFT solver for pfstmo_fattal02
	* updated: Added check for libexif
	* updated: improvements in pfstmo_ferradans11 to produce brighter results
	* updated: Updated some manual pages

pfstools 2.0.1 <07.01.2015>
	* added: new tone-mapping operator "pfstmo_ferradans11" (thanks to Sira)

pfstools 2.0.0 <06.12.2014>
	* added: Starting from pfstools 2.0.0, pfstmo and pfscalibration are included in the pfstools source package
	* changed: support for automake dropped in favour of cmake
	* fixed: pfsinpfm and pfsoutpfm now handle both big and little endian files correctly
	* fixed: pfsinpfm and pfsoutpfm do not flip images (top-bottom)
	* added: new tone-mapping operator: pfstmo_mai11
	* changed: Thoroughly updated README files
	* fixed: several issues with octave scripts (compatibility with newer Octave releases), in particular pfs_write_rgb

pfstools 1.9.0 <17.09.2014>
	* added: CMake build system (to replace automake in future releases)
	* changed: pfsinrgbe & pfsoutrgbe do not correct for Radiance WHITE_EFFICACY by default. Note that this makes absolute values stored in .hdr files using earlier pfstools version smaller then they should be. Check manual pages for pfsinrgbe.
	* added: pfsretime - simple frame duplication
	* added: pfscolortransform - color calibration
	* updated: pfsglview contains many improvement and multithreading (thanks to Oliver Barth)
	* updated: pfsview is multithreaded if openmp present
	* fixed: pfsindcraw uses sRGB rather than the native RGB color space. This should improve colors in HDR images merged from RAW.
	* fixed: bugs in pfsintiff
	* fixed: Building OSX app bundle for pfsview disabled by default (use --enable-osxapp to enable again)

pfstools 1.8.5 <21.10.2011>
	* fixed: pfsintiff normalizes 16-bit files to 0-1 range to maintain compatibility with pfscalibration and the rest of pfstools
	* fixed: when reading TIFF files, pfsin falls back to pfsinimgmagick if pfsintiff not found
	* fixed: pfsout tries first pfsoutimgmagic when writing TIFF images


src/fileformat/pfsoutyuv.cpp
 /**
 * @brief Write files in Radiance's RGBE format
 * 
 * This file is a part of PFSTOOLS package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Grzegorz Krawczyk, <krawczyk@mpi-sb.mpg.de>
 * @author Rafal Mantiuk, <mantiuk@mpi-sb.mpg.de>
 *
 * $Id: pfsoutrgbe.cpp,v 1.3 2014/06/17 21:57:09 rafm Exp $
 */

#include <config.h>
#include <string.h>

#include <cstdlib>

#include <iostream>
#include <cmath>
#include <getopt.h>
#include <pfs.h>
#include <pfsutils.cpp>

#define PROG_NAME "pfsoutyuv"

template<typename T>
T clip3(T x, T y, T z){
  if(z<x){
    return x;
  }
  if(z>y){
    return y;
  }
  return z;
}

class MetaDataFrameFileIterator : pfs::FrameFileIteratorImpl
{
  
  unsigned int width;
  unsigned int height;
  unsigned int bitdepth;
  unsigned int fps;
  const char * colorSpace;
  const char * chromaFormat;
  //Used for YUV output - we'll override this from FrameFileIterator - because we only want one YUV file writing to - and the rest of the class is useful
  
  public:
  MetaDataFrameFileIterator( int &argc, char* argv[], const char *fopenMode,
    const char *fileNamePrefix, FILE *stdinout,
    const char *optstring, const struct option *getopt_long, unsigned int width, unsigned int height, unsigned int bitdepth, const char * colorSpace, const char * chromaFormat, int fps) : 
    pfs::FrameFileIteratorImpl(argc, argv, fopenMode, fileNamePrefix, stdinout, optstring, getopt_long), bitdepth(bitdepth), colorSpace(colorSpace), width(width), height(height), chromaFormat(chromaFormat), fps(fps) 
  {
     
  }

  pfs::FrameFile getNextFrameFile(){
    if(!currentPattern->isPattern){
      strcpy(fileName, currentPattern->pattern);
    }
    else{ 
      char metaData[100]; //reasonably safe?
      
      if(fps>0){
        sprintf(metaData, "%dx%d_%dfps_%db_%s_%s", width, height, fps, bitdepth, colorSpace, chromaFormat);
      }
      else{ 
        sprintf(metaData, "%dx%d_%db_%s_%s", width, height, bitdepth, colorSpace, chromaFormat);
      }

      sprintf(fileName,currentPattern->pattern, metaData);
    
    }
    //grab prefix 
    FILE *fh;
    if( currentPattern->stdinout != NULL )
      fh = currentPattern->stdinout;
    else
      fh = fopen( fileName, fopenMode );
        if( fh != NULL ){
         return pfs::FrameFile( fh, fileName );
        }
        else{
          throw pfs::Exception("Cound't open file for saving");
        }
        return pfs::FrameFile(NULL, NULL);
    }
  
  void closeFrameFile(pfs::FrameFile ff){
    pfs::FrameFileIteratorImpl::closeFrameFile(ff);
  }
};

class FilterableArray2D : public pfs::Array2D {
  Array2D * base;
  public:
    FilterableArray2D(pfs::Array2D * in){
      base = in;
    } 

    int getCols() const{
      return base->getCols();
    }
    int getRows() const{
      return base->getRows();
    }
    
    float& operator()( int col, int row ){
      if(col < 0 || col >= base->getCols() || row < 0 || row >= base->getRows()){
        return *(new float(0));
      }
      else{
        return (*base)(col,row);
      }
    }

    const float& operator()( int col, int row ) const{
      if(col < 0 || col >= base->getCols() || row < 0 || row >= base->getRows()){
        return *(new float(0));
      }
      else{
        return (*base)(col,row);
      }
    }
    
    float& operator()( int index ){
      return (*base)(index);
    }
    
    const float& operator()( int index ) const{
      return (*base)(index); 
    }
};

class QuietException 
{
};

enum DownsampleFilter{f0, f1};


void printHelp()
{
  std::cerr << PROG_NAME " [--verbose] [--quiet] [--bitdepth] [--colorspace] [--downsample-filter] [--chroma-format]" 
            << std::endl
            << "See man page for more information." << std::endl;
}

void downscale_444_to_420(pfs::Array2D * channel, DownsampleFilter method){
  float weightsF0[] = {0.125, 0.75, 0.125}; //weights for f0 filter method
  float weightsF1[] = {0.25, 0.5, 0.25}; //ditto for f1
  float * weights;
  switch(method){
    case f0 : weights = weightsF0; 
    case f1 : weights = weightsF1;
  }
  //downscales a single channel
  //do rows first
  
  FilterableArray2D chan(channel);
  for (int i = 0; i<chan.getCols(); i++){
    for(int j = 0; j<chan.getRows(); j+=2){
      chan(i,j) = weights[0] * chan(i, j-1) + weights[1] * chan(i,j) + weights[2]* chan(i, j+1); 
    }
  }
  //do columns second
  for( int j = 0; j<chan.getRows(); j+=2){
    for( int i = 0; i<chan.getCols(); i+=2){
      chan(i,j) = weights[0] * chan(i-1, j) + weights[1]* chan(i,j) + weights[2]* chan(i+1, j);
    }
  }
  //the downsampled pixels are now stored in the even cells of channel
}

class YUVWriter {
  
  FILE * fh;
  int bitdepth;
  bool downscale_chroma;
  DownsampleFilter downsampleFilter;
  public:
  YUVWriter(FILE * fh, int bitdepth, bool downscale_chroma, DownsampleFilter downsampleFilter) : fh(fh), bitdepth(bitdepth), downscale_chroma(downscale_chroma), downsampleFilter(downsampleFilter){  
  }
  

  template<typename T>
  void writeYuvChannel(pfs::Array2D * in, int width, int height, int stride, float scale, float offset){
  for( int j = 0; j<height; j++){ 
    T * line_buf = new T[width];

    for (int i = 0; i<width; i++){
      line_buf[i] = clip3(0, (1<<bitdepth) - 1,
                  (int) round((*in)(i<<stride % width, j<<stride % height) * scale + offset));
    }
    fwrite(line_buf, sizeof(T), width, fh);
    delete[] line_buf;
  }
  //performs floating point to 10 bit conversion
  }

  template<typename T>
  void writeYuvImage(pfs::Array2D *R, pfs::Array2D *G, pfs::Array2D *B){
  //writes the entire image to fh R - Luminance, G - Cr, B - Cb
  unsigned int width_chroma = R->getCols();
  unsigned int height_chroma = R->getRows();
  unsigned int step_chroma = 0; 
  if(downscale_chroma){
    width_chroma = R->getCols()/2;
    height_chroma = R->getRows()/2;
    step_chroma = 1;
    downscale_444_to_420(G, downsampleFilter);
    downscale_444_to_420(B, downsampleFilter);
  }
  writeYuvChannel<T>(R, R->getCols(), R->getRows(), 0, 219.0f*(1<<(bitdepth-8)), 1<<(bitdepth -4));
  writeYuvChannel<T>(G, width_chroma, height_chroma, step_chroma, 224.0f*(1<<(bitdepth-8)), 1<<(bitdepth -1));
  writeYuvChannel<T>(B, width_chroma, height_chroma, step_chroma, 224.0f*(1<<(bitdepth-8)), 1<<(bitdepth -1));
  }
};

void writeFrames( int argc, char* argv[] )
{
  pfs::DOMIO pfsio;

  bool verbose = false;
  int bit_depth = 10;
  unsigned int width; 
  unsigned int height;
  unsigned int fps = 0;
  bool quiet = false;
  bool downscale_chroma = true;
  pfs::ColorSpace opt_colorspace = pfs::CS_INVALID;
  DownsampleFilter downsampleFilter = f0;
  bool srgb_input = false;  // Input is in sRGB colour space if true, RGB otherwise
  bool opt_srgb_input = false;  
  bool opt_linear_input = false;
  
  
  // Parse command line parameters
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "bitdepth", required_argument, NULL, 'b' },
    { "colourspace", required_argument, NULL, 'c' },
    { "quiet", no_argument, NULL, 'q' },
    { "downsample-filter", required_argument, NULL, 'a'},
    { "chroma-format", required_argument, NULL, 'f'},
    { "srgb-input", no_argument, NULL, 's'},
    { "linear-input", no_argument, NULL, 'l'},
    { NULL, 0, NULL, 0 } };
  static const char optstring[] = "hb:vqc:f:a:sl";
    
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, optstring, cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'b':
      bit_depth = strtol( optarg, NULL, 10 );
      break;
    case 'a':
      if( strcmp( optarg, "f0") == 0){
        downsampleFilter = f0;}
      else if( strcmp( optarg, "f1") == 0){
              downsampleFilter = f1;}
      break;
    case 'f':
       if( strcmp( optarg, "444") == 0){
         downscale_chroma = false;
       }
       else if ( strcasecmp( optarg, "420") == 0){
         downscale_chroma = true;
         
       }
       else{
        throw pfs::Exception( "Unrecognised chroma-sampling format" );
       }
       break;
  case 'c':
    if( !strcasecmp(optarg, "pq2020" )) {
      opt_colorspace = pfs::CS_PQYCbCr2020;
    } else if( !strcasecmp(optarg, "bt709" )) {
      opt_colorspace = pfs::CS_YCbCr709;
    } else if( !strcasecmp(optarg, "hlg2020" )) {
      opt_colorspace = pfs::CS_HLGYCbCr2020;
    } else{
      throw pfs::Exception( "Unrecognized colorspace name" );
    }
    break;
    case 'q':
      quiet = true;
      break;
    case 's':
      opt_srgb_input = true;
      break;
    case 'l':
      opt_linear_input = true;
      break;
      
  case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }


  //we need to open this here to grab metadata for the filename :)
  pfs::Frame *frame = pfsio.readFrame( stdin );  
  pfs::Channel *fullDimensionChannel;
  fullDimensionChannel = frame->getChannel("X");    
  if( fullDimensionChannel == NULL ) {       // No color
    throw pfs::Exception( "Missing X channel in the PFS stream" );
  }
  width = fullDimensionChannel->getCols();
  height = fullDimensionChannel->getRows();
  
  const char* luminanceTag = frame->getTags()->getString("LUMINANCE");
  const char* fpsTag = frame->getTags()->getString("FPS");
  if( opt_colorspace == pfs::CS_INVALID ) {
    // Try to guess the right colour space
    if( luminanceTag!=NULL && 0==strcmp(luminanceTag,"DISPLAY") ) {      
      opt_colorspace = pfs::CS_YCbCr709;
    } else if( luminanceTag!=NULL && 0==strcmp(luminanceTag,"ABSOLUTE") )  {
        opt_colorspace = pfs::CS_PQYCbCr2020;
    } else
      throw pfs::Exception( "Color space must be specified" );
  }

  const char * colorSpaceStr;
  switch( opt_colorspace ) {
  case pfs::CS_YCbCr709:
    colorSpaceStr = "bt709";
    break;    
  case pfs::CS_PQYCbCr2020:
    colorSpaceStr = "pq2020";
    break;    
  case pfs::CS_HLGYCbCr2020:
    colorSpaceStr = "hlg2020";
    break;    
  default:
    throw pfs::Exception( "Unrecognized color space" );    
  }

  if(fpsTag != NULL){
    VERBOSE_STR << "FPS Tag: " << fpsTag << std::endl;
    fps = atoi(fpsTag);
  }

  if( opt_srgb_input ) {    
    srgb_input = true;
  } else if( opt_linear_input ) {
    srgb_input = false;    
  } else {
    if( opt_colorspace == pfs::CS_YCbCr709 ) 
      srgb_input = true;    
  }
  
    if( !srgb_input && luminanceTag!=NULL && 0==strcmp(luminanceTag,"DISPLAY") ) {      
      std::cerr << "pfsoutyuv: Warning - assuming linear (HDR) input while the stream may be gamma-corrected (LDR)." << std::endl;
    }
    if( srgb_input && luminanceTag!=NULL && 0==strcmp(luminanceTag,"ABSOLUTE") ) {      
      std::cerr << "pfsoutyuv: Warning - assuming gamma-corrected (LDR) input while the stream may be linear (HDR)." << std::endl;
    }
    if( luminanceTag!=NULL && 0==strcmp(luminanceTag,"RELATIVE") ) {      
      std::cerr << "pfsoutyuv: Warning - relative HDR pixel values detected. Absolute values should be provided." << std::endl;
    }
    
  
  if( srgb_input )
    VERBOSE_STR << "Input data in sRGB (LDR) space" << std::endl;
  else
    VERBOSE_STR << "Input data in absolute linear (HDR) space" << std::endl;
  
  MetaDataFrameFileIterator it(argc, argv, "wb", NULL, stdout, optstring, cmdLineOptions, width,
                                height, bit_depth, colorSpaceStr, downscale_chroma ? "420" : "444", fps);
  
  pfs::FrameFile ff = it.getNextFrameFile();
  if( ff.fh == NULL ) {
    throw pfs::Exception("Couldn't open file for writing");
  } else {
      VERBOSE_STR << "Saving file: " << ff.fileName << std::endl << width << "x" << height << ", Bit-depth: " << bit_depth << ", Colorspace: " << getColorspaceString(opt_colorspace)  << ", " << (downscale_chroma ? "420, " : "444, ") << std::endl;
  }

  while(true){
    YUVWriter writer( ff.fh, bit_depth, downscale_chroma, downsampleFilter);

    pfs::Channel *X, *Y, *Z;
    frame->getXYZChannels( X, Y, Z );    
    if( X == NULL || Y == NULL || Z == NULL) {       // No color
      throw pfs::Exception( "Missing X, Y, Z channels in the PFS stream" );
    }
    if( srgb_input ) {
      pfs::transformColorSpace( pfs::CS_XYZ, X, Y, Z, pfs::CS_RGB, X, Y, Z );
      pfs::transformColorSpace( pfs::CS_SRGB, X, Y, Z, opt_colorspace, X, Y, Z );      
    } else{
      pfs::transformColorSpace( pfs::CS_XYZ, X, Y, Z, opt_colorspace, X, Y, Z );
    }
    if( bit_depth <= 8){
      writer.writeYuvImage<unsigned char>( X, Y, Z );
    }
    else{
      writer.writeYuvImage<unsigned short>( X, Y, Z );
    }

    pfsio.freeFrame( frame );

    frame = pfsio.readFrame( stdin );
    if( frame == NULL ) {
      break; // No more frames
    }
  }
  it.closeFrameFile( ff );
}


int main( int argc, char* argv[] )
{
  try {
    writeFrames( argc, argv );
  }
  catch( pfs::Exception ex ) {
    std::cerr << PROG_NAME << " error: " << ex.getMessage() << std::endl;
    return EXIT_FAILURE;
  }        
  catch( QuietException  ex ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}
